File: rust/linux_abs/src/raw_socket.rs
#![allow(non_camel_case_types)]
#![allow(dead_code)]

struct socket {

}

impl socket {
    pub fn create_raw() -> i32 {
        unsafe {
            libc::socket(libc::AF_PACKET, libc::SOCK_RAW, libc::ETH_P_ALL)
        }
    }

    pub fn close(fd : i32) {
        unsafe { libc::close(fd); }
    }
}

File: python3/raw_socket.py
import socket
from pkt_buffer import *

ETH_P_ALL = 3

class netos_raw_socket:
    def __init__(self, ifname):
        self.fd = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.htons(ETH_P_ALL))
        self.fd.bind((ifname, 0))

    def recv(self):
        return self.fd.recv(65535)

    def __exit__(self):
        self.fd.close()

if __name__ == "__main__":
    nr = netos_raw_socket("wlp4s0")
    while True:
        rx_buf = nr.recv()
        print("rx len: " + str(len(rx_buf)))


